### PR TITLE
Python: fix: filter non-assistant messages from AgentResponse in WorkflowAgent

### LINKED ISSUE #4261: Python: [Bug]: _convert_workflow_event_to_agent_response_updates emits user input again
### Description

When agent responds to user query, it appends the user input with assistant response when using streaming response along with workflow.as_agent()
# Sample Non Streaming response
```
-> Ask a question (or 'exit' to quit): hi
[Response]:
hiHello! Welcome to our school. I'm the Principal here. How can I help you today? If you have questions about mathematics, physics, or social science, I have excellent teachers who c
an assist you. 
                                                                                                                                       
-> Ask a question (or 'exit' to quit): a=3, b=5, what is a-b

[Response]:
hiHello! Welcome to our school. I'm the Principal here. How can I help you today? If you have questions about mathematics, physics, or social science, I have excellent teachers who c
an assist you.a=3, b=5, what is a-bI need to find a - b when a = 3 and b = 5.                                                                                                         
**Solution:**
a - b = 3 - 5 = **-2**The Maths Teacher has answered your question. When a = 3 and b = 5, a - b = -2. Is there anything else you'd like to learn about?

-> Ask a question (or 'exit' to quit): exit
Exiting.
```

# Sample Streaming response
```
-> Ask a question (or 'exit' to quit): hi

[Principal]:

hi
Hello! Welcome to our school. I'm the Principal here. How can I help you today? If you have any questions about mathematics, physics, or social science, I'd be happy to connect you w
ith the appropriate teacher.                                                                                                                                                          
-> Ask a question (or 'exit' to quit): a=3, b=5, what is a-b+a

[Principal]:

hi
Hello! Welcome to our school. I'm the Principal here. How can I help you today? If you have any questions about mathematics, physics, or social science, I'd be happy to connect you w
ith the appropriate teacher.                                                                                                                                                          a=3, b=5, what is a-b+a

[Maths Teacher]:

I need to find the value of a - b + a, where a = 3 and b = 5.

Let me substitute the values:

a - b + a = 3 - 5 + 3

Now I'll calculate step by step:
- 3 - 5 = -2
- -2 + 3 = **1**

The answer is **1**.

[Principal]:

The Maths Teacher has answered your question. The value of a - b + a, where a = 3 and b = 5, is 1. Let me know if you have any other questions!

-> Ask a question (or 'exit' to quit): exit
Exiting.

```

### Code Sample

```markdown
from agent_framework import Agent
from agent_framework_anthropic import AnthropicClient
from anthropic import AsyncAnthropicFoundry
from agent_framework_orchestrations import GroupChatBuilder

client = AnthropicClient(
    anthropic_client=AsyncAnthropicFoundry(timeout=120),
    model_id="claude-opus-4-5",
)

maths_agent = Agent(
    client=client,
    instructions="You are a Maths teacher. Answer the user's question about maths. If you don't know the answer, say you don't know.",
    name="Maths Teacher",
    description=("Maths expert. Answers questions about mathematics and provides explanations."),
    default_options={"max_tokens": 4096},
)

physics_agent = Agent(
    client=client,
    instructions="You are a Physics teacher. Answer the user's question about physics. If you don't know the answer, say you don't know.",
    name="Physics Teacher",
    description=("Physics expert. Answers questions about physics and provides explanations."),
    default_options={"max_tokens": 4096},
)

social_science_agent = Agent(
    client=client,
    instructions="You are a Social Science teacher. Answer the user's question about social science. If you don't know the answer, say you don't know.",
    name="Social Science Teacher",
    description=(
        "Social Science expert. Answers questions about social science and provides explanations."
    ),
    default_options={"max_tokens": 4096},
)

principal_agent = Agent(
    client=client,
    instructions=(
        "You are a Principal of a school. Your job is to answer the user's question by delegating to your teachers: the Maths Teacher, Physics Teacher and Social Science Teacher. "
        "If the question is about mathematics, delegate to the Maths Teacher. If the question is about physics, delegate to the Physics Teacher. If the question is about social science, delegate to the Social Science Teacher. "
        "If you don't know which teacher to delegate to, say you don't know."
    ),
    name="Principal",
    description=(
        "Principal agent that delegates questions to subject-specific teacher agents based on the topic of the question."
    ),
    default_options={"max_tokens": 4096},
)

workflow = GroupChatBuilder(
    orchestrator_agent=principal_agent,
    participants=[maths_agent, physics_agent, social_science_agent],
    max_rounds=10,
).build()

workflow_agent = workflow.as_agent(name="School Principal Agent")


async def chat_nonstream():
    while True:
        message = input("\nAsk a question (or 'exit' to quit): ")
        if message in ["exit", "quit"]:
            print("Exiting.")
            break
        else:
            response = await workflow_agent.run(message)
            print(f"\n[Response]:\n{response.text}")


async def chat_stream():

    while True:
        message = input("\nAsk a question (or 'exit' to quit): ")
        current_agent: str | None = None
        if message in ["exit", "quit"]:
            print("Exiting.")
            break
        else:
            async for chunk in workflow_agent.run(message, stream=True):
                # print(chunk.__dict__)
                author_name = getattr(chunk, "author_name", None)

                if author_name and author_name != current_agent:
                    current_agent = author_name
                    print(f"\n[{current_agent}]:\n")
                if chunk.text:
                    print(chunk.text, end="\n", flush=True)


if __name__ == "__main__":
    import asyncio

    # Streaming version with agent transition markers
    asyncio.run(chat_stream())
    # Non-streaming version
    # asyncio.run(chat_nonstream())
```

### Error Messages / Stack Traces

```markdown
No error messages
```

### Package Versions

agent-framework==1.0.0rc1

### Python Version

python=3.11

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by LEDazzio01**
## Summary

Fixes #4261

When a workflow (e.g., GroupChat) emits an `AgentResponse` containing the full conversation history (user + assistant messages), `WorkflowAgent` was converting **all** messages to output — including user messages. This caused user inputs to be re-emitted as part of the assistant response, with the problem compounding across successive turns as the conversation history grew.

## Root Cause

In `_workflows/_agent.py`, both the streaming path (`_convert_workflow_event_to_agent_response_updates`) and the non-streaming path (`_convert_workflow_events_to_agent_response`) iterated over **all** messages in `AgentResponse.messages` without filtering by role. When GroupChat-style orchestrators emit their full conversation history as an `AgentResponse`, this meant user messages were surfaced as assistant output.

## Fix

Filter `AgentResponse.messages` to only include messages with `role="assistant"` before converting them to output:

**Streaming path** (`_convert_workflow_event_to_agent_response_updates`):
```python
for msg in data.messages:
    if msg.role != "assistant":
        continue
    updates.append(...)
```

**Non-streaming path** (`_convert_workflow_events_to_agent_response`):
```python
if isinstance(data, AgentResponse):
    assistant_messages = [msg for msg in data.messages if msg.role == "assistant"]
    messages.extend(assistant_messages)
```

## Tests Added

Added `TestWorkflowAgentUserInputFiltering` class with 4 test cases:
- **Streaming**: Verifies only assistant messages from a mixed-role `AgentResponse` appear in streaming output
- **Non-streaming**: Same verification for non-streaming path
- **All-assistant (regression)**: Verifies no change in behavior when `AgentResponse` only contains assistant messages
- **Empty after filtering**: Verifies correct behavior when `AgentResponse` contains only user messages

## Scope

This fix targets only the `AgentResponse` branches. The `Message` and `list[Message]` branches are intentionally left unchanged, as those represent explicit workflow output where the executor has control over what it yields (and existing tests validate this behavior).

### CODE REVIEW

**Review by LEDazzio01 — COMMENTED** [review]
Thanks for the review! Both comments raise the same valid point about the `list[Message]` and `Message` branches not being filtered.

**Our rationale for scoping the fix to `AgentResponse` only:**

The `list[Message]` and `Message` branches represent **explicit executor output** — the developer directly controls what they yield via `ctx.yield_output()` or `ctx.send_message()`. Silently filtering by role in those branches could break intentional behavior (e.g., an executor that deliberately emits a user-role message as part of its output).

In contrast, `AgentResponse` is a higher-level wrapper that commonly contains **full conversation history** (as GroupChat orchestrators do), where user messages are incidental rather than intentional output.

The issue reporter's repro (#4261) uses `GroupChatBuilder` which emits `AgentResponse` via `AgentExecutor`, so the fix directly addresses the reported behavior. If GroupChat also has a code path that yields `list[Message]` with the full conversation, that's arguably a bug in the GroupChat orchestrator itself (it should yield only new assistant messages), not something WorkflowAgent should silently fix.

Happy to extend the filter if maintainers prefer a more defensive approach — just wanted to call out the tradeoff.

File: python/packages/core/tests/workflow/test_workflow_agent.py [hunk 1]
```diff
@@ -1296,3 +1296,142 @@ def test_merge_updates_function_result_no_matching_call(self):
...
+
+        # Should only have assistant messages (3 out of 5)
+        assert len(updates) == 3, f"Expected 3 assistant updates, got {len(updates)}"
+
+        # Verify all updates are assistant role
+        for update in updates:
+            assert update.role == "assistant", f"Expected role='assistant', got role='{update.role}'"
+
+        # Verify the content is correct
+        texts = [u.text for u in updates]
```
Comment (moonbox3) [inline python/packages/core/tests/workflow/test_workflow_agent.py:1355]: Consider also asserting `update.author_name` for each update to verify the production code's `author_name=msg.author_name` mapping is preserved through filtering. Currently only `role` and `text` are verified on streaming updates.

```suggestion
        # Verify all updates are assistant role
        for update in updates:
            assert update.role == "assistant", f"Expected role='assistant', got role='{update.role}'"

        # Verify author_name is propagated correctly
        assert updates[0].author_name == "Principal"
        assert updates[1].author_name == "Maths Teacher"
        assert updates[2].author_name == "Principal"
```

File: python/packages/core/tests/workflow/test_workflow_agent.py [hunk 2]
```diff
@@ -1296,3 +1296,142 @@ def test_merge_updates_function_result_no_matching_call(self):
...
+
+        workflow = WorkflowBuilder(start_executor=user_only_executor).build()
+        agent = workflow.as_agent("user-only-agent")
+
+        updates: list[AgentResponseUpdate] = []
+        async for update in agent.run("test", stream=True):
+            updates.append(update)
+
+        # No assistant messages means no updates
+        assert len(updates) == 0
```
Comment (moonbox3) [inline python/packages/core/tests/workflow/test_workflow_agent.py:1442]: Consider adding a test for non-assistant, non-user roles (e.g., 'system' or 'tool') to explicitly document that the filter drops those as well. This would prevent a future regression if someone expects system messages to pass through.

Comment (LEDazzio01) [reply python/packages/core/tests/workflow/test_workflow_agent.py:1442]: Copy all feedback - I'll address tonight.

**Comment by LEDazzio01** [comment]
All review feedback has been addressed in commit d6ba0f9:

**Already in previous commit:**
- ✅ `author_name` assertions added to `test_streaming_agent_response_filters_user_messages`
- ✅ Code comments clarifying intentional exclusion of system and tool roles (both streaming and non-streaming paths)

**New in this commit:**
- ✅ `test_streaming_agent_response_filters_system_and_tool_roles` — verifies that `system` and `tool` role messages are filtered out in both streaming and non-streaming paths, documenting the allowlist behavior explicitly
- ✅ `test_non_streaming_agent_response_empty_after_filtering` — non-streaming counterpart for the empty-after-filtering edge case

Thanks for the thorough review @moonbox3!

**Comment by LEDazzio01** [comment]
Quick status update — I addressed all of @moonbox3's feedback (adding `author_name` assertions, a test case for non-assistant/non-user roles, and clarifying the code comment about intentional exclusion of system/tool roles). However, it looks like the push didn't go through — the latest commit on the branch still predates the review. I'll re-push the fixes today.

**Comment by LEDazzio01** [comment]
Correction on my last comment — I just double-checked the branch and all 3 fixes are actually **already in the latest commit** (`d6ba0f9`):

1. ✅ `author_name` assertions added to `test_streaming_agent_response_filters_user_messages`
2. ✅ New `test_streaming_agent_response_filters_system_and_tool_roles` test added — documents that the allowlist intentionally drops system/tool roles
3. ✅ New `test_non_streaming_agent_response_empty_after_filtering` test added — non-streaming counterpart

The code comment in the new system/tool test docstring also clarifies the intentional allowlist behavior. No additional push needed — all feedback has been addressed. 🙌

**Comment by markwallace-microsoft** [comment]
<a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/README.md">Python Test Coverage</a><details><summary>Python Test Coverage Report • </summary><table><tr><th>File</th><th>Stmts</th><th>Miss</th><th>Cover</th><th>Missing</th></tr><tbody><tr><td colspan="5"><b>packages/core/agent_framework/_workflows</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py">_agent.py</a></td><td>362</td><td>70</td><td>80%</td><td><a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L66">66</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L74-L80">74&ndash;80</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L116-L117">116&ndash;117</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L207">207</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L253">253</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L255">255</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L318">318</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L320">320</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L374-L375">374&ndash;375</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L381-L382">381&ndash;382</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L388">388</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L390">390</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L395">395</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L455-L456">455&ndash;456</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L465">465</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L472">472</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L503">503</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L536-L538">536&ndash;538</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L540">540</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L542">542</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L544">544</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L549">549</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L554">554</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L607">607</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L637">637</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L654">654</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L693-L696">693&ndash;696</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L702">702</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L708">708</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L712-L713">712&ndash;713</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L716-L722">716&ndash;722</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L726-L727">726&ndash;727</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L735">735</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L796">796</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L803">803</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L809-L810">809&ndash;810</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L821">821</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L853">853</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L860">860</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L881">881</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L890">890</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L894">894</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L896-L898">896&ndash;898</a>, <a href="https://github.com/microsoft/agent-framework/blob/d6ba0f9e87ebde7909e36025f34f76e4ca644a98/packages/core/agent_framework/_workflows/_agent.py#L905">905</a></td></tr><tr><td><b>TOTAL</b></td><td><b>22225</b></td><td><b>2758</b></td><td><b>87%</b></td><td>&nbsp;</td></tr></tbody></table></details>

## Python Unit Test Overview
| Tests | Skipped | Failures | Errors | Time |
| ----- | ------- | -------- | -------- | ------------------ |
| 4724 | 247 :zzz: | 0 :x: | 0 :fire: | 1m 20s :stopwatch: |

**Comment by LEDazzio01** [comment]
@moonbox3 All your feedback has been addressed in the latest commits:

1. ✅ **`author_name` assertions**: Added to `test_streaming_agent_response_filters_user_messages`:
   ```python
   assert updates[0].author_name == "Principal"
   assert updates[1].author_name == "Maths Teacher"
   assert updates[2].author_name == "Principal"
   ```

2. ✅ **Non-assistant/non-user role test**: Added `test_streaming_agent_response_filters_system_and_tool_roles` — verifies that `system` and `tool` roles are also filtered out by the allowlist filter, and documents this behavior explicitly to prevent regressions. Covers both streaming and non-streaming paths.

3. ✅ **Empty-after-filtering test**: Added `test_non_streaming_agent_response_empty_after_filtering` as a non-streaming counterpart to the existing streaming empty test.

Ready for re-review. 🙏

**Comment by LEDazzio01** [comment]
All feedback from the automated review has been addressed in commits `dc51ea8` and `d6ba0f9`:

1. ✅ **`author_name` assertions** — Added `author_name` verification to the streaming filter test (`updates[0].author_name == "Principal"`, etc.)
2. ✅ **System/tool role test** — Added `test_streaming_agent_response_filters_system_and_tool_roles` covering both streaming and non-streaming paths with `"system"` and `"tool"` role messages, documenting that the allowlist filter drops all non-assistant roles intentionally.
3. ✅ **Non-streaming empty edge case** — Added `test_non_streaming_agent_response_empty_after_filtering` verifying empty messages list when all messages are non-assistant.

Ready for re-review. 🙏

**Comment by moonbox3** [comment]
Sorry for the delay on this. #4482 is a more comprehensive approach to solving this issue. It introduces an `OrchestrationComplete` event type that wraps the full conversation at orchestration termination, cleanly separating internal workflow state (needed for passing context between agents) from user-facing output. This avoids the need for role-based filtering at the `WorkflowAgent` boundary. Closing in favor of #4482.
